src/esp/io/JsonBuiltinTypes.h

```diff
@@ -0,0 +1,285 @@
+// before all toJsonValue/fromJsonValue definitions, and they are defined
+// later in JsonBuiltinTypes.hpp. The quirky ordering is to avoid some compile
+// errors.
+
+template <typename T>
+void addMember(JsonGenericValue& value,
+               rapidjson::GenericStringRef<char> name,
+               const T& obj,
+               JsonAllocator& allocator);
+
+template <typename T>
+bool readMember(const JsonGenericValue& value, const char* name, T& x);
+
+/**
+ * @brief Fallback implementation for fromJsonValue to produce a runtime error
+ * for types that haven't implemented fromJsonValue.
+ */
+template <typename T>
+bool fromJsonValue(CORRADE_UNUSED const JsonGenericValue& obj,
+                   CORRADE_UNUSED T& val) {
+  // If you've already implemented fromJsonValue for your type and you're still
+  // hitting this, the underlying issue might be the ordering among json helper
+  // definitions.
+  LOG(ERROR) << "Unsupported type. Aborting. You need to implement "
+                "fromJsonValue for typeid(T).name() = "
+             << typeid(T).name() << ".";
+  return false;
+}
+
+/**
+ * @brief Fallback implementation for toJsonValue to produce a runtime error for
+ * types that haven't implemented toJsonValue.
+ */
+template <typename T>
+JsonGenericValue toJsonValue(const T&, JsonAllocator&) {
+  // If you've already implemented toJsonValue for your type and you're still
+  // hitting this, the underlying issue might be the ordering among json helper
+  // definitions.
+  LOG(ERROR) << "Unsupported type. Aborting. You need to implement toJsonValue "
+                "for typeid(T).name() = "
+             << typeid(T).name() << ".";
+  return JsonGenericValue(rapidjson::kObjectType);
+}
+
+// toJsonValue wrappers for the 7 rapidjson builtin types. A JsonGenericValue
+// can be directly constructed from the builtin types.
+
+inline JsonGenericValue toJsonValue(bool x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(int x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(unsigned x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(int64_t x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(uint64_t x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(double x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+inline JsonGenericValue toJsonValue(float x, JsonAllocator&) {
+  return JsonGenericValue(x);
+}
+
+// fromJsonValue wrappers for the 7 rapidjson builtin types
+
+/**
+ * @brief Populate passed @p val with value. Returns whether successfully
+ * populated, or not. Logs an error if inappropriate type.
+ *
+ * @param obj string tag to look for in json doc
+ * @param val destination value to be populated
+ * @return whether successful or not
+ */
+inline bool fromJsonValue(const JsonGenericValue& obj, bool& val) {
```

Comment:
The implementation and documentation for these was ported from jsonToVal functions in json.h.